Add all instructions as a feature

boomanaiden154 opened the pull request:
**Note:** This patch is currently based against the `added config options for vocab generation` patch, so a lot of the changes in this PR are related to that. The only relevant changes for adding the instruction features are the creation of the new model in `compiler_opt/rl/regalloc/config.py` at the top of the file and the adding of this model as a preprocessing layer for the `instructions_and_mapping` feature.

The current input tensors is a 34x300 matrix with the first row being the instruction opcodes and the next 33 rows being a binary matrix which maps each live range to its instructions. Eg:
```
op_code,op_code,op_code,op_code...
1            ,1            ,1            ,0           ...
0            ,0            ,0            ,1           ...
...
```

Currently the model performs the following steps on the input tensor:
1. Split the matrix into two parts, the first one being the instruction opcodes and the next one being the binary mapping matrix
2. Get embeddings for the instruction opcodes
3. Multiply the instruction embeddings by the binary matrix to get an output 33x16 matrix

Adding in a sparse optimizations on the matrix multiplication would probably be optimal, but tensorflow agents passes in the batch size as an additional dimension on the tensors passed to each preprocessing layer, and while the normal matrix multiplication algorithm supports having a batch size dimension, for some reason when enabling sparse optimizations, the function doesn't support having the batch size dimension. I'm planning on seeing how trivial this is to fix upstream and maybe writing a patch or just submitting an issue.

I'll try and post other potential models here that could potentially improve the performance of the model (since the matrix multiplication is extremely expensive) while still passing very similar/the same data to the model.

boomanaiden154 committed:
Added regalloc all instructions config

boomanaiden154 commented:
@yundiqian Here are the changes for the initial model plan. I'll try and get out an implementation of the idea that I mentioned on slack soon. Sorry for the delayed timeline. Validation that all the code worked as expected took longer than expected.

yundiqian commented on compiler_opt/rl/regalloc/config.py:
```diff
@@ -14,23 +14,68 @@
+  return 17716
+
+def get_num_instructions():
+  return 300
```
where does this number come from? what if the 'affected #lines of code' is larger or smaller?

boomanaiden154 replied:
It's just a number that I settled on after some experimentation. It seems to be able to encompass most of the instructions for every instance of the problem while not being too massively large. If the number of opcodes being sent is lower, the input gets zero-padded, and if the number of opcodes sent is larger, the data is just truncated.

yundiqian commented on compiler_opt/rl/regalloc/config.py:
```diff
@@ -14,23 +14,68 @@
+  def call(self, inputs):
+    # The first row is the instructions, and all rows afterwards compose the binary
+    # mapping matrix
+    instruction_opcodes = tf.slice(inputs, begin=[0, 0, 0], size=[-1, 1, self.instruction_count])
```
so this is the op code for the 300 lines of code

boomanaiden154 replied:
Yes. I'm passing everything in as a single `int64_t` matrix so that I can grab everything in one preprocessing layer and then splitting it up within the model (with this line grabbing the instruction opcodes).

yundiqian commented on compiler_opt/rl/regalloc/config.py:
```diff
@@ -14,23 +14,68 @@
+    # mapping matrix
+    instruction_opcodes = tf.slice(inputs, begin=[0, 0, 0], size=[-1, 1, self.instruction_count])
+    instruction_opcodes = tf.reshape(instruction_opcodes, [-1, self.instruction_count])
+    binary_mapping_matrix = tf.slice(inputs, begin=[0, 1, 0], size=[-1, self.register_count, self.instruction_count])
```
this is a very sparse matrix about which live range contains which line of code

boomanaiden154 replied:
Yes. The whole input matrix (opcodes + mappings) is sparse, so it might even be best to just take a sparse tensor as input and convert the instruction opcodes vector to dense once it is extracted, but that point is moot due to the unavailability of a `sparse_dense_mat_mul` function that works with a batch size dimension.

yundiqian commented on compiler_opt/rl/regalloc/config.py:
```diff
@@ -14,23 +14,68 @@
+    binary_mapping_matrix_casted = tf.cast(binary_mapping_matrix, tf.float32)
+    instruction_embeddings = self.embedding_layer(instruction_opcodes)
+
+    matrix_product = tf.linalg.matmul(binary_mapping_matrix_casted, instruction_embeddings)
```
this is the bag-of-word aggregation I mentioned

boomanaiden154 replied:
Yes. Just grabbing embeddings and then doing the matrix multiplication. There are a couple other options that I want to look at for the implementation of the bag of words aggregation that would in theory be more efficient than matrix multiplication. I'm going to try and get some code written up for that today.

boomanaiden154 commented:
```python
class process_instruction_features(tf.keras.Model):

  def __init__(self, opcode_count, register_count, instructions_per_live_range_count, embedding_dimensions=16):
    super().__init__()
    self.opcode_count = opcode_count
    self.register_count = register_count
    self.instructions_per_live_range_count = instructions_per_live_range_count
    self.embedding_dimensions = embedding_dimensions
    self.embedding_layer = tf.keras.layers.Embedding(self.opcode_count,
                                                     self.embedding_dimensions,
                                                     input_length=self.instructions_per_live_range_count * self.register_count)

  def get_config(self):
    return {
      "opcode_count": self.opcode_count,
      "register_count": self.register_count,
      "instructions_per_live_range_count": self.instructions_per_live_range_count,
      "embedding_dimensions": self.embedding_dimensions
    }

  @classmethod
  def from_config(cls, config):
    return cls(**config)

  def call(self, inputs):
    # inputs is a batch_size x register_count x instructions_per_live_range tensor containing opcodes.
    instruction_embeddings = self.embedding_layer(inputs)
    # we now have a batch_size, register_count, instructions_per_live_range, embedding_dimensions shaped tensor
    # sum along the instruction_count dimension
    output_tensor = tf.reduce_sum(instruction_embeddings, 2)
    return output_tensor
```
@yundiqian This was the other method I was describing on slack, where instead of taking a linear sequence of instructions and a mapping matrix, all of the instructions are just sent over as opcodes in a `num_registers * num_instructions_per_live_range` shaped tensor. Then an embedding layer is run on all of them and there is a reduction in the form of a sum/average. This should theoretically be faster than doing a full matrix multiplication, but I do want to make sure I'm not missing something.

yundiqian commented:
This is a working solution for simple bag-of-word with only op code information about instructions. However, the issue with this representation is that:
1. information is duplicated X time (X == how many live ranges one instructions touches); it will be more severe if we want to include more per-instruction features other than op code in the future, e.g., frequency of this instruction.
2. the information of how different live ranges interference with each other gets lost, so we can hardly do anything more than the bag-of-word here

> ```python
> class process_instruction_features(tf.keras.Model):
> 
>   def __init__(self, opcode_count, register_count, instructions_per_live_range_count, embedding_dimensions=16):
>     super().__init__()
>     self.opcode_count = opcode_count
>     self.register_count = register_count
>     self.instructions_per_live_range_count = instructions_per_live_range_count
>     self.embedding_dimensions = embedding_dimensions
>     self.embedding_layer = tf.keras.layers.Embedding(self.opcode_count,
>                                                      self.embedding_dimensions,
>                                                      input_length=self.instructions_per_live_range_count * self.register_count)
> 
>   def get_config(self):
>     return {
>       "opcode_count": self.opcode_count,
>       "register_count": self.register_count,
>       "instructions_per_live_range_count": self.instructions_per_live_range_count,
>       "embedding_dimensions": self.embedding_dimensions
>     }
> 
>   @classmethod
>   def from_config(cls, config):
>     return cls(**config)
> 
>   def call(self, inputs):
>     # inputs is a batch_size x register_count x instructions_per_live_range tensor containing opcodes.
>     instruction_embeddings = self.embedding_layer(inputs)
>     # we now have a batch_size, register_count, instructions_per_live_range, embedding_dimensions shaped tensor
>     # sum along the instruction_count dimension
>     output_tensor = tf.reduce_sum(instruction_embeddings, 2)
>     return output_tensor
> ```
> 
> @yundiqian This was the other method I was describing on slack, where instead of taking a linear sequence of instructions and a mapping matrix, all of the instructions are just sent over as opcodes in a `num_registers * num_instructions_per_live_range` shaped tensor. Then an embedding layer is run on all of them and there is a reduction in the form of a sum/average. This should theoretically be faster than doing a full matrix multiplication, but I do want to make sure I'm not missing something.

yundiqian commented:
for the high-dimension sparse dense matrix multiplication, i found this github issue about it: https://github.com/tensorflow/tensorflow/issues/9210

the most making-sense solution is mentions is https://github.com/tensorflow/tensorflow/issues/9210#issuecomment-850780779, but the author complains that it cannot calculate gradient. i tried a bit on it and it seems to be working well, including gradient computation:

```
import tensorflow as tf

st1 = tf.sparse.SparseTensor(indices=[[0, 0, 2], [1, 1, 1]], values=[10., 20.], dense_shape=[2, 3, 3])
dt1 = tf.Variable([[[1,2,3],[4,5,6],[7,8,9]],[[1,2,3],[4,5,6],[7,8,9]]], dtype=tf.float32)

def sparse_dense_matmult_batch(sp_a, b):
    return tf.map_fn(
        lambda x: tf.sparse.sparse_dense_matmul(x[0], x[1]),
        elems=(sp_a, b), fn_output_signature=tf.float32
    )

for _ in range(20):
  with tf.GradientTape() as tape:
    a = sparse_dense_matmult_batch(st1, dt1)
    b = tf.reduce_sum(a)
    loss = b * b
    tf.print(loss)

  grad = tape.gradient(loss, dt1)
  dt1.assign_sub(0.0001 * grad)
```

boomanaiden154 commented:
So I've done some more testing in regards to sparse matrices. I currently can't get the entire matrix to get passed to the `process_instruction_features` model as a sparse matrix (at least fully functionally, it does work, but the policy saver breaks) due to the fact that a `SparseTensorSpec` doesn't have a name attribute. I'll work on playing around with this a little bit more though because it is absolutely possible that I am just missing something. I also tried just converting the `binary_mapping_matrix` to a sparse matrix after it has been sliced and then passing it into the `sparse_dense_matmult_batch` function that you sent. However,  when doing it like this I'm seeing significantly reduced number of steps per second running behavioral cloning. I'm not totally sure why this is currently, but I'd hypothesize it's because `map_fn` isn't taking advantage of parallelism as efficiently as the non-sparse optimized `linalg.matmul` function, or that the conversion of the dense matrix into a sparse one is just taking up a significant amount of time. I'll continue to play around with it though and see if I can get something working.

yundiqian commented:
ic, then i'm wondering how about just abandoning the sparse tensor concept. so the input becomes:
1. 300 * op_code instructions as before; later converted to 300 * 16 embedding table
2. 33 * 5 (or probably variable length, can start with a fixed number, e.g., 5) --- 5 being the #instructions this live range def/use, so for example it is [2, 5, 7, 9, 10] for live range 1, means live range 1 use/def instructions at line 2, 5, 7, 9, 10. Then you can use tf.gather to have 33 * 5 * 16 embedding table, then you do the tf.reduce_sum() to make it 33 * 16

in this way, it does the same thing as the sparse matrix multiplication, but does not contain sparse matrix

boomanaiden154 commented:
That should be pretty easily doable. I'll work soon on getting the necessary work done on the LLVM side to extract those features and then the ml-compiler-opt side should be pretty easy. I'll probably experiment with using all instructions that the live range encompasses (as I'm doing in this PR) as well as just use/def instructions as both should be fairly straightforward. It is also doesn't look too hard to add a name attribute to the `SparseTensorSpec` class, so I might experiment with that tonight and see if anything comes out of it. If it does, there might still be some issues to resolve (eg tensorflow version compatibility, seems to have already been resolved though), upstreaming, and then just waiting for a release cycle. Shouldn't hurt to at least try though.

boomanaiden154 commented:
Just keeping track of progress here. I hacked on `SparseTensor` a little bit last night and was able to pretty easily add a `name` attribute. However, just adding a name attribute doesn't push anything a whole lot farther as it doesn't get around the issue that all the logging on the LLVM side doesn't support sparse tensors. This is just for behavioral cloning, completely ignoring the whole problem of getting a saved model with a sparse tensor input working on the LLVM side (this would again require some rework on that side). Maybe something to look into supporting at some point, but definitely not a direction that is easy enough to pursue at this current time. Leaving the code changes in the `all-features-tooling-sparse-tensor` branch.

yundiqian commented:
ic, i guess then we can just use the [2, 5, 7, 9, 10] index representations. the only issue with it is that it does not support variable lengths easily, but i guess it's fine for now.

mtrofin committed:
[infra] Introduce a simple local-process worker pool manager (#33)

[infra] Introduce a simple local-process worker pool manager

This is meant as an in-place replacement of the current functionality,
but using a stateful worker object abstraction analogous to what Dask
supports. For local workloads, this implementation is faster than Dask's
LocalCluster (which may be due to misconfiguration). Because it's very
simple and introduces no new dependencies, even if LocalCluster
performance could be improved, the implementation would still be useful
for debugging.

The worker abstraction allows implementers specify a list of methods
that should be executed promptly on the server side, analogous to Dask's
`separate_thread=False` concept. For our purposes, we can use it to
implement full cancelation of work on the server side.

This patch introduces the worker manager, a subsequent patch will enable
its use in the rest of the codebase.

boomanaiden154 committed:
Benchmarking utilities (#47)

boomanaiden154 committed:
Refactor dev dependencies to separate requirements file (#51)

mtrofin committed:
Use LocalWorkerManager (#48)

This change switches to using LocalWorkerManager. The main change to the compilation runner code is that cancellation is now handled 'all-out' - we just cancel all jobs on the worker.

Issue #31

mtrofin committed:
Fix spelling in benchmark_chromium.py

mtrofin committed:
Fix spelling in benchmarking.md

Northbadge committed:
Fix generate_default_trace (#54)

Regression introduced in 09b41a8 (missing delete flags)

Northbadge committed:
Fix generate_default_trace.py (#56)

Regression introduced in 32a5c1c
Runners became un-pickle-able due to Runners containing a WorkerCancellationManager which contained a threading.Lock()

boomanaiden154 committed:
Refactored train_eval to use num_modules from gin configs (#53)

Northbadge committed:
[NFC] yapf auto reformatting (#55)

yundiqian committed:
test for generate_default_trace (#50)

Also:
- Handle flags appearing as duplicate to pytest

The flags appear duplicate because pytest loads all the test modules,
and for tests for tools like generate_default_trace or extract_ir, which
define flags, some "popular" flag names may be thus duplicate.

- dropping tf_agents system_multiprocessing because it doesn't work
well with pytest.

Northbadge committed:
Switch file_paths to ModuleSpec (#57)

Preliminary patch to refactoring file_paths tuple usage completely.

Allows distributed thinlto corpora to be trained on

mtrofin committed:
Added type annotation and comment. (#60)

boomanaiden154 committed:
Add yapf to CI and dev dependencies (#59)

Since we are now using yapf to format the repository, it makes sense to
have yapf in the development requirements. This commit also adds a yapf
format check to the CI to ensure that all new code changes are properly
formatted.

mtrofin committed:
Make generate_default_trace fail-fast when hitting a bug (#61)

If the worker process hits an unrecoverable error, we stop processing
instead of blocking indefinitely.

mtrofin committed:
Fix formatting issue introduced in PR#61

boomanaiden154 committed:
Added dockerfile with base dependencies (#49)

Northbadge committed:
Add support for extracting lld-thinlto corpora (#58)

boomanaiden154 committed:
Make the tensorboard data interval adjustable (#62)

Currently, all of the tensorboard variables get updated every
single iteration. This is quite costly in terms of performance. I'm
seeing about a 25% performance improvement by just running this once
every 100 iterations instead of running it every iteration. The data
showing up in tensorboard will definitely be more sparse, but it should
be good enough to see the overall trends that we're looking for.

Northbadge committed:
Move corpus related functions to corpus.py (#63)

yundiqian committed:
move regalloc_network to regalloc/regalloc_network (#64)

* move regalloc_network to regalloc/regalloc_network

* add back import in bc.gin

Northbadge committed:
Replace module_paths with corpus_description.json (#65)

* Replace module_paths with metadata.json

- This adds 'global_command_override' functionality for lld-thinlto corpora, exposed in metadata.json
- Setting the above in the metadata.json file will default the runners to using that as the base command, rather than looking for a .cmd file

* Rename to corpus_description.json

* Fix pytype error

* Make UNSPECIFIED a module constant

* Move UNSPECIFIED to constant.py. Resolved comments

mtrofin committed:
Type APIs working with policies (#66)

This clarifies we use TFAgents rather than raw saved models, and helps
with readability and discoverability (the latter in an IDE in
particular)

Northbadge committed:
Minor fixes (#67)

* Minor fixes

- the '-cc1' flag is required for commands extracted from .cmd files- Sometimes it's not in the .cmd file (Not sure how I managed it, but it happened)
- disabled yapf formatting on data_reader's dataset builder
- gracefully handle when a compile_command doesn't have the -o option

mtrofin committed:
CI: parallelize tasks, and support local 'act' runs (#69)

most work is now parallelized. If running locally with act, only 3.8 is run - this is because typically that's where most of the errors (like typing) will be found (e.g. stuff that's supported post-3.8 won't break unless tried on 3.8)

On act, pip package caching doesn't work (via actions/setup-python); instead, run act with -r. After the first run, it should complete in ~30-40s.

Northbadge committed:
Fix metric logging (#72)

- Decouples the summary intervals, so one is not reliant on the other

- Switches to modulo instead of remembering the last step number, this is more robust, especially if training is restarted at some step which is not a multiple of the set intervals (which previously would cause no data to be written)

yundiqian committed:
move #!/bin/bash down to make copybara happy. (#74)

afaul committed:
corrected libtensorflow dependency for demo (#73)

Northbadge committed:
Set max workers for LocalWorkerManager (#75)

- Prevents attempting to spawn 32 * min(32, cores) number of clangs (python errors out)
- max_worker count is adjustable via `pool_threads` parameter in LocalWorkerPool constructor

yundiqian committed:
import modules instead of importing packages (#76)

petrhosek committed:
Script to build TFLite and its dependencies (#79)

This script builds TFLite together with all its dependencies and
generates a CMake cache file that can be used from other projects.

boomanaiden154 committed:
Add development mode dockerfile and info on use (#78)

mtrofin committed:
LocalWorker: Fix nr threads to 1 (#81)

mtrofin committed:
Enable building tflite on 'dev*'  buildbots (#82)

Also bumping Ruy hash to capture last build-related fix

mtrofin committed:
Add few additional settings to build_tflite

Northbadge committed:
Fix worker shutdown (#83)

If process was killed while its output queue is accessed, the queue would be left broken, causing msg_pump to not die, and .join() to never unblock.

Also, fix a misc. indentation bug on a test case

kshiteejm committed:
Update to Contributions (#87)

* updates to existing contributions page and linking it in the main README

* remove trailining whitespaces

* clarified guidelines in contributions

Northbadge committed:
Tidy up local_worker_manager (#88)

- Add a get_context()
- Fix a signature
- Update module docstring

Northbadge committed:
Add one more get_context() (#89)

Northbadge committed:
Promote locals in tests to module scope (#90)

Local-scoped classes won't be pickle-able following an internal change.

boomanaiden154 committed:
Refactor add/delete flags options to problem_configuration (#84)

This patch moves the add/delete flags from a gin config option to the
problem configuration python object and makes generate_default_trace
grab the flags from there.

boomanaiden154 committed:
Fix formatting for CI (#92)

boomanaiden154 committed:
Prevent TF from logging info messages during data collection (#93)

Currently, whenever a TF model gets loaded in by LLVM in the
development mode advisors, TF outputs a slew of info messages
to STDOUT that aren't particularly useful. This patch gets
rid of those messages by setting an environment variable when
running the compilation commands that only enables warnings
and errors to be printed to the console. This makes it much
easier to see progress/results in several of the scripts within
this repository such as train_locally.py and
generate_default_trace.py when grabbing performance data for a
policy.

Northbadge committed:
Minor cosmetic changes (#99)

Part of #96

Northbadge committed:
Add Corpus abstraction & sorting w/ size (#77)

* Add Corpus abstraction & sorting w/ size

- Rather than pass around a list of module_specs,pass around a Corpus object instead
- built in filter and sample method
- sample will sort by size descending, with the goal of optimizing compile order

* change .size -> __len__

* Default to unbiased sampling, add sampler option

* Add separate constructor for testing

* Make sampler not select repeats

* replace 'corp' with 'cps'

* Switch to optimized algorithm

* resolve comments

Northbadge committed:
Replace _compile_fn with compile_fn (#107)

Will be used directly by #96

Northbadge committed:
Resolve potential deadlock in local_worker_manager (#108)

- acquire lock -> future.set_result -> done_callback() -> ... -> _Stub. __getattr__ -> remote_call -> fetch lock -> deadlocked
- Releasing the lock earlier resolves this.

Northbadge committed:
Add load balancer (#106)

Adds a buffered load balancer, which by default maintains at least 2 tasks assigned to each worker.

Closes #91

mtrofin committed:
Fix some incorrectly specified flags

Northbadge committed:
Add doc for thinlto corpus extraction (#111)

yundiqian committed:
add tensorboard instruction to demo.md (#105)

Northbadge committed:
Add pause/resume/context to workers (#101)

* Add pause/resume/context to workers

- Allows a user to start/stop processes at will, via OS signals SIGSTOP and SIGCONT.
- Allows a user to bind processes to specific CPUs.
- Allows local_worker_pool to be used outside of a context manager
- Switch workers to be Protocol based, so Workers are effectively duck-typed (i.e. anything that has the required methods passes as a Worker)

Part of #96

Northbadge committed:
Fix failing test (#112)

mtrofin committed:
Update tensorflow to 2.10-rc1 to enable building tflite.

mtrofin committed:
Fix arguments-renamed warning introduced by TF 2.10

mtrofin committed:
Revert move to TF 2.10-rc1

This reverts commits 6ec41b9323f7e3304ef7451b5194f6f1382d9cf6
and 1804d498b52fc86bd41c0538e730f40abd8afdff.

A runtime error is introduced, needs more investigation. See
https://github.com/google/ml-compiler-opt/runs/7932290113?check_suite_focus=true

boomanaiden154 committed:
Moved to multi input preprocessing layer

boomanaiden154 committed:
Fixed problem config and added flags to the problem config
